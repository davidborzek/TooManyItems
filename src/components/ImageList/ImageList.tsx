import React from 'react';
import {
  Image,
  Text,
  View,
  FlatList,
  StyleSheet,
  RefreshControl,
  TouchableOpacity,
} from 'react-native';
import { Ionicons } from '@expo/vector-icons';

export type ImageListItem = {
  name: string;
  image?: string | null;
};

type Props<T extends ImageListItem> = {
  items: T[];
  refreshing?: boolean;
  onRefresh?: () => void;
  onPress?: (item: T) => void;
  onLongPress?: (item: T) => void;
  ListEmptyComponent?:
    | React.ComponentType<any>
    | React.ReactElement
    | null
    | undefined;
  imagePlaceholderIcon?: React.ComponentProps<typeof Ionicons>['name'];
};

const styles = StyleSheet.create({
  item: {
    flexDirection: 'row',
    alignItems: 'center',
    padding: 10,
    backgroundColor: 'light-gray',
  },
  image: {
    width: 48,
    height: 48,
    borderRadius: 2,
  },
  imagePlaceholder: {
    justifyContent: 'center',
    alignItems: 'center',
    borderWidth: 2,
    borderColor: '#c1c1c1',
  },
  containerName: {
    fontSize: 18,
    marginLeft: 16,
  },
});

export default function ImageList<T extends ImageListItem>({
  items,
  onRefresh,
  refreshing,
  onPress,
  onLongPress,
  ListEmptyComponent,
  imagePlaceholderIcon,
}: Props<T>) {
  return (
    <FlatList
      data={items}
      renderItem={({ item }) => {
        const callOnPress = () => {
          if (onPress) {
            onPress(item);
          }
        };
        return (
          <TouchableOpacity
            style={styles.item}
            onPress={callOnPress}
            onLongPress={() => {
              onLongPress && onLongPress(item);
            }}
            delayLongPress={200}
          >
            {item.image ? (
              <Image source={{ uri: item.image }} style={styles.image} />
            ) : (
              <View style={[styles.image, styles.imagePlaceholder]}>
                <Ionicons
                  name={imagePlaceholderIcon || 'cube'}
                  size={24}
                  color="#c1c1c1"
                />
              </View>
            )}
            <Text style={styles.containerName}>{item.name}</Text>
          </TouchableOpacity>
        );
      }}
      refreshControl={
        refreshing != undefined ? (
          <RefreshControl refreshing={refreshing} onRefresh={onRefresh} />
        ) : undefined
      }
      ListEmptyComponent={ListEmptyComponent}
      contentContainerStyle={{ flex: items.length === 0 ? 1 : undefined }}
    />
  );
}
