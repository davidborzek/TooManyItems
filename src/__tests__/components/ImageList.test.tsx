import ImageList, { ImageListItem } from '../../components/ImageList/ImageList';

import React from 'react';
import { Text } from 'react-native';
import renderer from 'react-test-renderer';

const items: ImageListItem[] = [
  {
    image: 'Some image',
    name: 'Some name',
  },
];

it('ImageList renders correctly with required props and no items', () => {
  const tree = renderer.create(<ImageList items={[]} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('ImageList renders correctly with required props and items', () => {
  const tree = renderer.create(<ImageList items={items} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('ImageList renders correctly refreshing', () => {
  const tree = renderer.create(<ImageList items={items} refreshing />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('ImageList renders correctly with all props', () => {
  const tree = renderer
    .create(
      <ImageList
        items={items}
        ListEmptyComponent={<Text>Empty State</Text>}
        onLongPress={() => ''}
        onPress={() => ''}
        onRefresh={() => ''}
        imagePlaceholderIcon="add"
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
