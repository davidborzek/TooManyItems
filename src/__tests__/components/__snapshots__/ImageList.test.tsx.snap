// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageList renders correctly refreshing 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "flex": undefined,
    }
  }
  data={
    Array [
      Object {
        "image": "Some image",
        "name": "Some name",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  refreshControl={
    <RefreshControlMock
      refreshing={true}
    />
  }
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  viewabilityConfigCallbackPairs={Array []}
>
  <RCTRefreshControl />
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "light-gray",
            "flexDirection": "row",
            "opacity": 1,
            "padding": 10,
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "data:image/png;base64,Some image",
            }
          }
          style={
            Object {
              "borderRadius": 2,
              "height": 48,
              "width": 48,
            }
          }
        />
        <Text
          style={
            Object {
              "fontSize": 18,
              "marginLeft": 16,
            }
          }
        >
          Some name
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`ImageList renders correctly with all props 1`] = `
<RCTScrollView
  ListEmptyComponent={
    <Text>
      Empty State
    </Text>
  }
  contentContainerStyle={
    Object {
      "flex": undefined,
    }
  }
  data={
    Array [
      Object {
        "image": "Some image",
        "name": "Some name",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  viewabilityConfigCallbackPairs={Array []}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "light-gray",
            "flexDirection": "row",
            "opacity": 1,
            "padding": 10,
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "data:image/png;base64,Some image",
            }
          }
          style={
            Object {
              "borderRadius": 2,
              "height": 48,
              "width": 48,
            }
          }
        />
        <Text
          style={
            Object {
              "fontSize": 18,
              "marginLeft": 16,
            }
          }
        >
          Some name
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`ImageList renders correctly with required props and items 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "flex": undefined,
    }
  }
  data={
    Array [
      Object {
        "image": "Some image",
        "name": "Some name",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  viewabilityConfigCallbackPairs={Array []}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "light-gray",
            "flexDirection": "row",
            "opacity": 1,
            "padding": 10,
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "data:image/png;base64,Some image",
            }
          }
          style={
            Object {
              "borderRadius": 2,
              "height": 48,
              "width": 48,
            }
          }
        />
        <Text
          style={
            Object {
              "fontSize": 18,
              "marginLeft": 16,
            }
          }
        >
          Some name
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`ImageList renders correctly with required props and no items 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "flex": 1,
    }
  }
  data={Array []}
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  viewabilityConfigCallbackPairs={Array []}
>
  <View />
</RCTScrollView>
`;
